components/script/dom/globalscope.rs

```diff
@@ -604,6 +634,365 @@ impl GlobalScope {
             .insert(message_port_id.clone(), port_impl);
     }
 
+    /// Return a clone of a blob impl, as part of
+    /// https://html.spec.whatwg.org/multipage/#serialization-steps
+    pub fn clone_blob_impl(&self, blob_id: &BlobId) -> BlobImpl {
+        // TODO: in case a a file-based blob, we should be able to avoid reading byte
+        // and simply increment the file ref.
+        let bytes = self
+            .get_blob_bytes(blob_id)
+            .expect("To be able to read bytes for cloning a blob");
+        let type_string = self.get_blob_type_string(blob_id);
+        BlobImpl::new_from_bytes(bytes, type_string)
+    }
+
+    fn track_blob_impl(&self, blob_impl: BlobImpl) {
+        // Store the BlobImpl.
+        self.blobs
+            .borrow_mut()
+            .insert(blob_impl.blob_id(), blob_impl);
+    }
+
+    /// Start tracking blobs
+    pub fn track_blob(&self, dom_blob: &DomRoot<Blob>, blob_impl: BlobImpl) {
+        // Store the DOM object.
+        self.blob_tracker
+            .borrow_mut()
+            .insert(blob_impl.blob_id(), WeakRef::new(dom_blob));
+        self.track_blob_impl(blob_impl);
+    }
+
+    /// Start tracking files
```

Comment:
I've tried, it seems easier to keep two lists, because it's the `File` that contains a `Blob`, and while we could track that internal blob, it seems that once you've put it into a `DomRoot`(which seems necessary to track it), it's hard to take it out again to store it inside the file.